From bf5ea493bc43ba2f57ca25c8f624e2b4ba1e3347 Mon Sep 17 00:00:00 2001
From: Roland Haas <rhaas@illinois.edu>
Date: Wed, 23 Jan 2019 10:11:40 -0600
Subject: [PATCH] MemSpeed: use at most 1GB of memory per MPI rank for tests

this helps on workstations with lots of memory and is fine until the L3
cache is 1GB is size.
---
 MemSpeed/src/memspeed.cc | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/MemSpeed/src/memspeed.cc b/MemSpeed/src/memspeed.cc
index 4e0fd20..3e6d8d2 100644
--- a/MemSpeed/src/memspeed.cc
+++ b/MemSpeed/src/memspeed.cc
@@ -12,6 +12,7 @@
 #include <cstring>
 #include <string>
 #include <vector>
+#include <algorithm>
 
 using namespace std;
 
@@ -472,13 +473,15 @@ void calc_memsizes(int cache, ptrdiff_t &skip_memsize, ptrdiff_t &memsize) {
     break;
   case mem_local:
     skip_memsize = 0;
-    memsize = cache_info[cache].size / 4;
+    // use either 1/4 of main memory or 1GB for memtest, whichever is smaller
+    memsize = min(cache_info[cache].size / 4, ptrdiff_t(1024*1024*1024));
     break;
   case mem_global:
     assert(cache > 0);
     assert(cache_info[cache - 1].type == mem_local);
     skip_memsize = cache_info[cache - 1].size;
-    memsize = skip_memsize / 4;
+    // use either 1/4 of main memory or 1GB for memtest, whichever is smaller
+    memsize = min(skip_memsize / 4, ptrdiff_t(1024*1024*1024));
     assert(skip_memsize + memsize <= cache_info[cache].size * 3 / 4);
     break;
   }
-- 
2.20.1

